src/sources/statsd/mod.rs

```diff
@@ -50,75 +105,102 @@ impl SourceConfig for StatsdConfig {
-                                        None
-                                    }
-                                })
-                                .collect::<Vec<_>>();
-                            Some(stream::iter(metrics))
-                        }
-                        Err(error) => {
-                            emit!(StatsdSocketError::read(error));
-                            None
-                        }
-                    }
-                })
-                .flatten()
-                .forward(out.sink_compat())
-                .await;
-
-            info!("Finished sending");
-            Ok(())
+    let socket = UdpSocket::bind(&config.address)
+        .map_err(|error| emit!(StatsdSocketError::bind(error)))
+        .await?;
+
+    info!(
+        message = "Listening.",
+        addr = %config.address,
+        r#type = "udp"
+    );
+
+    let _ = UdpFramed::new(socket, BytesCodec::new())
+        .take_until(shutdown.compat())
+        .filter_map(|frame| {
+            future::ready(match frame {
+                Ok((bytes, _sock)) => {
+                    let packet = String::from_utf8_lossy(bytes.as_ref());
+                    let metrics = packet
+                        .lines()
+                        .filter_map(|line| match parse(line) {
+                            Ok(metric) => {
+                                emit!(StatsdEventReceived {
+                                    byte_size: line.len()
+                                });
+                                Some(Ok(Event::Metric(metric)))
+                            }
+                            Err(error) => {
+                                emit!(StatsdInvalidRecord { error, text: line });
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>();
+                    Some(stream::iter(metrics))
+                }
+                Err(error) => {
+                    emit!(StatsdSocketError::read(error));
+                    None
+                }
+            })
+        })
+        .flatten()
+        .forward(out.sink_compat())
+        .await;
+
+    Ok(())
+}
+
+#[derive(Clone)]
+struct StatsdTcpSource;
+
+impl TcpSource for StatsdTcpSource {
+    type Error = std::io::Error;
+    type Decoder = BytesDelimitedCodec;
+
+    fn decoder(&self) -> Self::Decoder {
+        BytesDelimitedCodec::new(b'\n')
+    }
+
+    fn build_event(&self, line: Bytes, _host: Bytes) -> Option<Event> {
+        let line = String::from_utf8_lossy(line.as_ref());
+        match parse(&line) {
+            Ok(metric) => {
+                emit!(StatsdEventReceived {
+                    byte_size: line.len()
+                });
+                Some(Event::Metric(metric))
+            }
+            Err(error) => {
+                emit!(StatsdInvalidRecord { error, text: &line });
+                None
+            }
         }
```

Comment:
This is a duplicate of the parse stanza in the UDP source above, consider adding a wrapper function.